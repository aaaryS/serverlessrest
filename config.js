module.exports = [{
  tableName: 'action_points',
  item: 'ActionPoint',
  attributes: [
    { name: 'description', type: 'String', required: true },
    { name: 'created_at', type: 'String' },
    { name: 'archived_at', type: 'String' },
    { name: 'due_date', type: 'String' },
    { name: 'sprint', type: 'String' },
    { name: 'state', type: 'String', required: true },
    { name: 'team', type: 'String' },
    { name: 'actions_taken', type: 'String' },
    { name: 'owners', type: '[String]' },
  ],
  queries: [
    { type: 'get' },
    { type: 'index' },
    {
      type: 'custom',
      name: 'actionPointsByTeam',
      attributes: [
        {
          name: 'team',
          type: 'String',
          required: true,
        },
      ],
      returnType: 'collection',
    },
  ],
  mutations: [
    {
      type: 'create',
      attributes: [
        { name: 'description', type: 'String!' },
        { name: 'state', type: 'String!' },
        { name: 'archived_at', type: 'String' },
        { name: 'due_date', type: 'String' },
        { name: 'sprint', type: 'String' },
        { name: 'team', type: 'String!' },
        { name: 'actions_taken', type: 'String' },
        { name: 'owners', type: '[String]' },
      ],
    },
    {
      type: 'update',
      attributes: [
        { name: 'state', type: 'String' },
        { name: 'actions_taken', type: 'String' },
        { name: 'archived_at', type: 'String' },
      ],
    },
    {
      type: 'delete',
    },
  ],
},
{
  item: 'SubTask',
  attributes: [
    { name: 'title', type: 'String', required: true },
  ],
},
{
  tableName: 'tasks',
  item: 'Task',
  attributes: [
    { name: 'title', type: 'String', required: true },
    { name: 'description', type: 'String', required: true },
  ],
  queries: [
    { type: 'get' },
    { type: 'index' },
  ],
  mutations: [
    {
      type: 'create',
      attributes: [
        { name: 'description', type: 'String!' },
        { name: 'title', type: 'String!' },
      ],
    },
    {
      type: 'update',
      attributes: [
        { name: 'description', type: 'String' },
        { name: 'title', type: 'String' },
      ],
    },
    {
      type: 'delete',
    },
  ],
}];
