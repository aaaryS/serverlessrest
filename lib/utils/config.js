import { readFileSync } from 'fs';
import parseArgs from 'minimist';

const normalizeConfig = config => config.map((_c) => {
  const c = { ..._c };
  if (!c.queries) c.queries = [];
  if (!c.mutations) c.mutations = [];
  if (!c.type) c.type = 'object';
  if (!c.auth) c.auth = false;

  return c;
});

const hasAuth = config => config.filter(table => table.auth).length > 0;

const loadConfig = () => {
  const configPath = parseArgs(process.argv).config;

  if (configPath) {
    const content = readFileSync(`${process.cwd()}/${configPath}`).toString();
    // eslint-disable-next-line no-eval
    global.eval(content.replace('module.exports = configExport', ''));
    return global.configExport;
  }

  // eslint-disable-next-line global-require
  return require('../../examples/complex/complex.config');
};

export default {
  normalizeConfig,
  loadConfig,
  hasAuth,
};
