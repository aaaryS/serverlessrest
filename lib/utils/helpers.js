export const lowerFirstLetter = str => str.replace(/^\w/, c => c.toLowerCase());

export const captialize = str => str.replace(/^\w/, c => c.toUpperCase());

export const delimiter = spaces => `\n${Array(spaces).fill(' ').join('')}`;
