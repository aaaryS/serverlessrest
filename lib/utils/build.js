import { delimiter } from './helpers';

// eslint-disable-next-line import/prefer-default-export
export const buildAttributes = (attributes, spaces = 0, withNewLines = true) => (
  attributes.map(attr => (
    `${attr.name}: ${attr.type}${attr.required ? '!' : ''}`
  )).join(delimiter(spaces, withNewLines))
);
