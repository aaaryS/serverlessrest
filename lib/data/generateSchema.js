import { makeExecutableSchema } from 'graphql-tools';
import { mergeResolvers, mergeTypes } from 'merge-graphql-schemas';

import generateType from './types/generateType';
import generateAuthTypes from './types/generateAuthTypes';
import generateResolver from './resolvers/generateResolver';
import generateAuthResolvers from './resolvers/generateAuthResolvers';

import { normalizeConfig } from './../utils/config';

const generateDefs = (_config) => {
  const authTypes = generateAuthTypes(_config);
  if (authTypes) {
    return mergeTypes([..._config.map(generateType), authTypes]);
  }

  return mergeTypes(_config.map(generateType));
};

const generateResolvers = (_config) => {
  const authResolvers = generateAuthResolvers(_config);
  if (authResolvers) {
    return mergeResolvers([..._config.map(generateResolver), authResolvers]);
  }

  return mergeResolvers(_config.map(generateResolver));
};

const generateSchema = (config) => {
  const normalizedConfig = normalizeConfig(config);

  const typeDefs = generateDefs(normalizedConfig);

  console.log(typeDefs)

  const resolvers = generateResolvers(normalizedConfig);

  return makeExecutableSchema({
    typeDefs,
    resolvers,
  });
};

export default generateSchema;
