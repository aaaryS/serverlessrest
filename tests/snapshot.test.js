import { generateDefs } from '../lib/data/generateSchema';
import { normalizeConfig } from '../lib/utils/config';

import basicConfig from '../examples/basic/basic.config';
import authConfig from '../examples/auth/auth.config';
import complexConfig from '../examples/complex/complex.config';

it('should match snapshot for basic config', () => {
  const normalizedConfig = normalizeConfig(basicConfig);
  const def = generateDefs(normalizedConfig);

  expect(def).toMatchSnapshot();
});

it('should match snapshot for auth config', () => {
  const normalizedConfig = normalizeConfig(authConfig);
  const def = generateDefs(normalizedConfig);

  expect(def).toMatchSnapshot();
});

it('should match snapshot for complex config', () => {
  const normalizedConfig = normalizeConfig(complexConfig);
  const def = generateDefs(normalizedConfig);

  expect(def).toMatchSnapshot();
});
