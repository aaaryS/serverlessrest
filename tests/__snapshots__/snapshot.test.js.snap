// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot for auth config 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

type Query {
  tasks: [Task]
  task(id: ID!): Task
  me: User
}

type Mutation {
  createTask(title: String!): Task
  deleteTask(id: ID!): Task
  signup(username: String!, email: String!, password: String!): String
  login(email: String!, password: String!): String
  me: Object
}

type Task {
  id: ID!
  title: String!
}

scalar Object

type User {
  id: Int!
  username: String!
  email: String!
  password: String!
}
"
`;

exports[`should match snapshot for basic config 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

type Query {
  tasks: [Task]
  task(id: ID!): Task
}

type Mutation {
  createTask(description: String!, priority: Int, title: String!, created_at: String, tags: [String]): Task
  updateTask(id: ID!, description: String, title: String, tags: [String]): Task
  deleteTask(id: ID!): Task
}

type Task {
  id: ID!
  title: String!
  priority: Int
  description: String!
  tags: [String]
  created_at: String
}
"
`;

exports[`should match snapshot for complex config 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

type Query {
  epics: [Epic]
  epic(id: ID!): Epic
  epicsBySprint(sprint: String!): [Epic]
  tasks: [Task]
  task(id: ID!): Task
  me: User
}

type Mutation {
  createEpic(description: String!, archived_at: String, due_date: String, sprint: String): Epic
  updateEpic(id: ID!, description: String, sprint: String): Epic
  deleteEpic(id: ID!): Epic
  createTask(description: String!, title: String!): Task
  updateTask(id: ID!, description: String, title: String, t_type: TaskType!): Task
  deleteTask(id: ID!): Task
  signup(username: String!, email: String!, password: String!): String
  login(email: String!, password: String!): String
  me: Object
}

type Epic {
  id: ID!
  description: String!
  created_at: String
  archived_at: String
  due_date: String
  sprint: String
}

enum TaskType {
  MAIN
  SUBTASK
  ARCHIVED
}

type Task {
  id: ID!
  title: String!
  description: String!
  t_type: TaskType
}

scalar Object

type User {
  id: Int!
  username: String!
  email: String!
  password: String!
}
"
`;
